### PR TITLE
OCPBUGSM-27416 Handle AAAA DNS records in Route53

### PR DESCRIPTION
**Pull request by empovit**
Create an AAAA record if the address is IPv6.

When checking if a record exists for a DNS name,
check for both A and AAAA records.

Refactor to hide DNS provider implementation in validations.

### CODE REVIEW

File: internal/dns/dns.go [hunk 1]
```diff
@@ -44,67 +45,98 @@ func NewDNSHandler(baseDNSDomains map[string]string, log logrus.FieldLogger) DNS
...
+func (h *handler) DeleteDNSRecordSets(ctx context.Context, cluster *common.Cluster) error {
+	log := logutil.FromContext(ctx, h.log)
+	if err := h.updateDNSRecordSet(log, cluster, h.deleteDNSRecord); err != nil {
+		return err
+	}
+	log.Infof("Successfully deleted DNS records for base domain: %s", cluster.BaseDNSDomain)
+	return nil
+}
 
+func (h *handler) updateDNSRecordSet(log logrus.FieldLogger, cluster *common.Cluster, updateRecordFunc func(logrus.FieldLogger, *DNSDomain, string, string) error) error {
```
Comment (danielerez) [inline internal/dns/dns.go:66]: Maybe worth soring the context/log in the handler so we won't have to pass the log to all internal functions here. Could simplify the signatures.

File: internal/dns/dns.go [hunk 2]
```diff
@@ -44,67 +45,98 @@ func NewDNSHandler(baseDNSDomains map[string]string, log logrus.FieldLogger) DNS
...
 		}
-		log.Infof("Successfully created DNS records for base domain: %s", cluster.BaseDNSDomain)
+	}
+	return nil
+}
+
+func (h *handler) getDNSProvider(domain *DNSDomain, ip string) dnsproviders.Provider {
+	switch domain.Provider {
+	case "route53":
+
```
Comment (danielerez) [inline internal/dns/dns.go:125]: nit: redundant empty line

File: internal/dns/dns.go (hunk 1 above)
Comment (empovit) [reply internal/dns/dns.go:66]: I'm not sure. IIUC, the handler is reused between requests, while the context (and log created from it) is per request.

Comment (danielerez) [reply internal/dns/dns.go:77]: Yeah I've meant storing it only when calling the public functions (CreateDNSRecordSets/DeleteDNSRecordSets) then reuse it in the helpers. Just if worth the effort.

File: internal/dns/dns.go [hunk 3]
```diff
@@ -141,14 +181,80 @@ func (h *handler) GetDNSDomain(clusterName, baseDNSDomainName string) (*DNSDomai
...
+func (f *defaultDNSProviderFactory) GetProvider(domain *DNSDomain) dnsproviders.Provider {
+	switch domain.Provider {
+	case "route53":
+		return dnsproviders.Route53{
+			HostedZoneID: domain.ID,
+			SharedCreds:  true,
+		}
+	}
+	f.log.Debugf("No suitable implementation for DNS provider %s", domain.Provider)
```
Comment (eranco74) [inline internal/dns/dns.go:258]: Why do we need this if there is only one provider (route53)?

Comment (empovit) [reply internal/dns/dns.go:258]: @danielerez?

Comment (danielerez) [reply internal/dns/dns.go:258]: The switch was added just for future expandability, i.e. support a new provider. Can be refactored though if not needed, but probably out of the scope of this PR.

**Pull request merged**